File: students/cbt/extract_physics_questions.py
#!/usr/bin/env python3
"""
Physics Question Extraction using VLM (Vision Language Model)
Extracts questions from PDF as images with accurate boundaries
"""

import os
import sys
import json
import base64
import re
from pathlib import Path
from pdf2image import convert_from_path
import requests

# Configuration
PDF_PATH = "/home/z/my-project/students/cbt/pdfs/physics.pdf"
OUTPUT_DIR = "/home/z/my-project/students/cbt/images/physics_questions"
VLM_API_URL = "https://api.zukijourney.com/v1/chat/completions"
API_KEY = "sk-zuki-7f8e9d2a1b3c4f5e6a7b8c9d0e1f2a3b"  # Replace with actual key

# Year configurations - pages are 0-indexed
YEAR_CONFIG = {
    "2010": {"start_page": 1, "end_page": 10, "num_questions": 50},  # Pages 2-11 in PDF
    "2011": {"start_page": 11, "end_page": 20, "num_questions": 50},  # Pages 12-21
    "2012": {"start_page": 21, "end_page": 30, "num_questions": 50},  # Pages 22-31
    "2013": {"start_page": 31, "end_page": 40, "num_questions": 50},  # Pages 32-41
    "2014": {"start_page": 41, "end_page": 50, "num_questions": 50},  # Pages 42-51
    "2015": {"start_page": 51, "end_page": 58, "num_questions": 40},  # Pages 52-59
    "2016": {"start_page": 59, "end_page": 66, "num_questions": 40},  # Pages 60-67
    "2017": {"start_page": 67, "end_page": 74, "num_questions": 40},  # Pages 68-75
    "2018": {"start_page": 75, "end_page": 82, "num_questions": 40},  # Pages 76-83
}

def encode_image_to_base64(image_path):
    """Convert image to base64 string"""
    with open(image_path, "rb") as f:
        return base64.b64encode(f.read()).decode("utf-8")

def analyze_page_with_vlm(image_path, year, page_num):
    """
    Use VLM to analyze a page and identify question boundaries.
    Returns a list of questions with their bounding boxes.
    """
    base64_image = encode_image_to_base64(image_path)
    
    prompt = f"""Analyze this JAMB Physics exam page from year {year}.

Your task: Identify ALL questions visible on this page and extract their bounding boxes.

For EACH question, identify:
1. The question number
2. The bounding box of the ENTIRE question (including any diagrams/figures) - but EXCLUDING the options (A, B, C, D)
3. Whether the question has a diagram

IMPORTANT RULES:
- Each question includes: the question number, question text, and any associated diagram
- Do NOT include the options (A,..., D) in the bounding box
- If a question spans multiple columns, include the full question
- If there's a diagram/figure, include it in the bounding box

Return a JSON array with this format:
[
  {{
    "question_num": 1,
    "bbox": [x1, y1, x2, y2],  // normalized 0-1000
    "has_diagram": true/false,
    "question_preview": "First 50 chars of question text..."
  }}
]

Only return the JSON array, no other text."""

    try:
        response = requests.post(
            VLM_API_URL,
            headers={
                "Authorization": f"Bearer {API_KEY}",
                "Content-Type": "application/json"
            },
            json={
                "model": "gpt-4o",
                "messages": [
                    {
                        "role": "user",
                        "content": [
                            {"type": "text", "text": prompt},
                            {"type": "image_url", "image_url": {"url": f"data:image/png;base64,{base64_image}"}}
                        ]
                    }
                ],
                "max_tokens": 2000
            },
            timeout=60
        )
        
        if response.status_code == 200:
            result = response.json()
            content = result["choices"][0]["message"]["content"]
            # Extract JSON from response
            json_match = re.search(r'\[.*\]', content, re.DOTALL)
            if json_match:
                return json.loads(json_match.group())
        else:
            print(f"VLM API error: {response.status_code} - {response.text}")
    except Exception as e:
        print(f"VLM analysis error: {e}")
    
    return []

def convert_pdf_to_images(pdf_path, output_dir, dpi=200):
    """Convert PDF pages to images"""
    print(f"Converting PDF to images at {dpi} DPI...")
    images = convert_from_path(pdf_path, dpi=dpi)
    
    os.makedirs(output_dir, exist_ok=True)
    
    image_paths = []
    for i, image in enumerate(images):
        image_path = os.path.join(output_dir, f"page_{i+1:03d}.png")
        image.save(image_path, "PNG")
        image_paths.append(image_path)
        print(f"  Saved page {i+1}/{len(images)}")
    
    return image_paths

def extract_question_image(page_image_path, bbox, output_path, page_width, page_height):
    """Extract a question region from a page image"""
    from PIL import Image
    
    # Convert normalized coordinates (0-1000) to pixel coordinates
    x1 = int(bbox[0] * page_width / 1000)
    y1 = int(bbox[1] * page_height / 1000)
    x2 = int(bbox[2] * page_width / 1000)
    y2 = int(bbox[3] * page_height / 1000)
    
    # Add small padding
    padding = 5
    x1 = max(0, x1 - padding)
    y1 = max(0, y1 - padding)
    x2 = min(page_width, x2 + padding)
    y2 = min(page_height, y2 + padding)
    
    # Open and crop
    with Image.open(page_image_path) as img:
        cropped = img.crop((x1, y1, x2, y2))
        cropped.save(output_path, "PNG")
    
    return output_path

def process_year(year, config, page_images, output_dir):
    """Process all questions for a specific year"""
    print(f"\n{'='*50}")
    print(f"Processing year {year}")
    print(f"{'='*50}")
    
    questions = []
    start_page = config["start_page"]
    end_page = config["end_page"]
    
    # Collect all questions from all pages of this year
    all_page_questions = []
    
    for page_idx in range(start_page, min(end_page + 1, len(page_images))):
        page_path = page_images[page_idx]
        print(f"\nAnalyzing page {page_idx + 1}...")
        
        # Get page dimensions
        from PIL import Image
        with Image.open(page_path) as img:
            page_width, page_height = img.size
        
        # Analyze with VLM
        vlm_results = analyze_page_with_vlm(page_path, year, page_idx)
        
        for q in vlm_results:
            q["page_idx"] = page_idx
            q["page_width"] = page_width
            q["page_height"] = page_height
            q["page_path"] = page_path
            all_page_questions.append(q)
    
    # Sort by question number
    all_page_questions.sort(key=lambda x: x["question_num"])
    
    # Extract each question image
    for q in all_page_questions:
        q_num = q["question_num"]
        output_path = os.path.join(output_dir, f"{year}_q{q_num}.png")
        
        try:
            extract_question_image(
                q["page_path"],
                q["bbox"],
                output_path,
                q["page_width"],
                q["page_height"]
            )
            
            questions.append({
                "num": q_num,
                "image": f"images/physics_questions/{year}_q{q_num}.png",
                "has_diagram": q.get("has_diagram", False),
                "page": q["page_idx"] + 1
            })
            print(f"  Extracted question {q_num}")
        except Exception as e:
            print(f"  Error extracting question {q_num}: {e}")
    
    return questions

def main():
    """Main extraction process"""
    print("="*60)
    print("Physics Question Extraction using VLM")
    print("="*60)
    
    # Create output directory
    os.makedirs(OUTPUT_DIR, exist_ok=True)
    
    # Convert PDF to images
    page_images = convert_pdf_to_images(PDF_PATH, OUTPUT_DIR, dpi=200)
    print(f"\nConverted {len(page_images)} pages")
    
    # Process each year
    all_questions = {}
    
    for year, config in YEAR_CONFIG.items():
        questions = process_year(year, config, page_images, OUTPUT_DIR)
        all_questions[year] = questions
        print(f"\nYear {year}: Extracted {len(questions)} questions")
    
    # Save metadata
    metadata_path = os.path.join(OUTPUT_DIR, "extraction_metadata.json")
    with open(metadata_path, "w") as f:
        json.dump(all_questions, f, indent=2)
    print(f"\nSaved metadata to {metadata_path}")
    
    # Summary
    print("\n" + "="*60)
    print("EXTRACTION SUMMARY")
    print("="*60)
    total = 0
    for year, questions in all_questions.items():
        count = len(questions)
        total += count
        expected = YEAR_CONFIG[year]["num_questions"]
        status = "OK" if count == expected else f"MISSING {expected - count}"
        print(f"  {year}: {count}/{expected} questions - {status}")
    print(f"\nTotal extracted: {total} questions")

if __name__ == "__main__":
    main()

File: scripts/extract-physics-questions.ts
import ZAI from 'z-ai-web-dev-sdk';
import fs from 'fs';
import path from 'path';
import sharp from 'sharp';

// Page mapping for Physics PDF (based on PDF structure)
// Year starts on specific pages
const YEAR_PAGE_MAP: { [year: string]: { start: number; end: number; count: number } } = {
  '2010': { start: 2, end: 11, count: 50 },  // Pages 2-11 (0-indexed: 1-10)
  '2011': { start: 12, end: 21, count: 50 }, // Pages 12-21
  '2012': { start: 22, end: 31, count: 50 }, // Pages 22-31
  '2013': { start: 32, end: 41, count: 50 }, // Pages 32-41
  '2014': { start: 42, end: 51, count: 50 }, // Pages 42-51
  '2015': { start: 52, end: 57, count: 40 }, // Pages 52-57
  '2016': { start: 58, end: 63, count: 40 }, // Pages 58-63
  '2017': { start: 64, end: 69, count: 40 }, // Pages 64-69
  '2018': { start: 70, end: 75, count: 40 }, // Pages 70-75
};

const PAGES_DIR = '/home/z/my-project/students/cbt/images/physics_pages';
const OUTPUT_DIR = '/home/z/my-project/students/cbt/images/physics_questions';

interface QuestionBoundary {
  questionNumber: number;
  topY: number;    // Percentage from top (0-100)
  bottomY: number; // Percentage from top (0-100)
  hasDiagram: boolean;
}

interface PageAnalysis {
  pageNumber: number;
  year: string;
  questions: QuestionBoundary[];
}

async function analyzePageWithVLM(zai: any, imagePath: string, pageNumber: number): Promise<QuestionBoundary[]> {
  const imageBuffer = fs.readFileSync(imagePath);
  const base64Image = imageBuffer.toString('base64');
  
  const prompt = `You are analyzing a page from a JAMB Physics past questions PDF.

This is page ${pageNumber} of the document. 

TASK: Identify ALL exam questions visible on this page and their vertical positions.

For EACH question found, provide:
1. The question NUMBER (e.g., 1, 2, 3, etc.)
2. The TOP position - percentage from the top of the page where the question starts (0% = very top, 100% = very bottom)
3. The BOTTOM position - percentage from the top where the question ends (including any diagram but EXCLUDING the options A, B, C, D)
4. Whether the question has a DIAGRAM (true/false)

IMPORTANT RULES:
- Each question should include its text and any diagram/figure
- DO NOT include the options (A., B., C., D.) in the boundary - stop before the options
- The bottom boundary should be just before "A." starts
- If a question continues from previous page or to next page, mark only what's visible
- Look carefully for question numbers like "1.", "2.", "3.", etc.
- Diagrams are usually images/figures associated with questions

Respond ONLY with a valid JSON array, no other text:
[
  {
    "questionNumber": 1,
    "topY": 5,
    "bottomY": 25,
    "hasDiagram": true
  },
  ...
]

If no exam questions are found on this page, return: []`;

  try {
    const response = await zai.chat.completions.createVision({
      messages: [
        {
          role: 'user',
          content: [
            { type: 'text', text: prompt },
            { type: 'image_url', image_url: { url: `data:image/png;base64,${base64Image}` } }
          ]
        }
      ],
      thinking: { type: 'disabled' }
    });

    const content = response.choices[0]?.message?.content || '[]';
    
    // Extract JSON from response
    let jsonStr = content;
    const jsonMatch = content.match(/\[[\s\S]*\]/);
    if (jsonMatch) {
      jsonStr = jsonMatch[0];
    }
    
    const questions: QuestionBoundary[] = JSON.parse(jsonStr);
    return questions;
  } catch (error) {
    console.error(`Error analyzing page ${pageNumber}:`, error);
    return [];
  }
}

async function extractQuestionFromPage(
  pageImagePath: string,
  question: QuestionBoundary,
  outputPath: string
): Promise<boolean> {
  try {
    const image = sharp(pageImagePath);
    const metadata = await image.metadata();
    
    if (!metadata.height || !metadata.width) {
      console.error('Could not get image metadata');
      return false;
    }
    
    const topPixel = Math.floor((question.topY / 100) * metadata.height);
    const bottomPixel = Math.floor((question.bottomY / 100) * metadata.height);
    const height = bottomPixel - topPixel;
    
    if (height <= 0) {
      console.error(`Invalid height for question: ${JSON.stringify(question)}`);
      return false;
    }
    
    // Extract the question region with small padding
    const padding = 5;
    const extractTop = Math.max(0, topPixel - padding);
    const extractHeight = Math.min(metadata.height - extractTop, height + padding * 2);
    
    await image
      .extract({ left: 0, top: extractTop, width: metadata.width, height: extractHeight })
      .toFile(outputPath);
    
    return true;
  } catch (error) {
    console.error(`Error extracting question:`, error);
    return false;
  }
}

async function main() {
  console.log('Starting Physics question extraction using VLM...\n');
  
  const zai = await ZAI.create();
  
  // Ensure output directory exists
  if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
  }
  
  const allExtractions: { [year: string]: { [num: number]: string } } = {};
  
  for (const [year, info] of Object.entries(YEAR_PAGE_MAP)) {
    console.log(`\n=== Processing ${year} (${info.count} questions) ===`);
    allExtractions[year] = {};
    
    // Process each page for this year
    for (let pageNum = info.start; pageNum <= info.end; pageNum++) {
      const pageImagePath = path.join(PAGES_DIR, `page-${String(pageNum).padStart(2, '0')}.png`);
      
      if (!fs.existsSync(pageImagePath)) {
        console.log(`Page ${pageNum} not found at ${pageImagePath}`);
        continue;
      }
      
      console.log(`Analyzing page ${pageNum}...`);
      
      const questions = await analyzePageWithVLM(zai, pageImagePath, pageNum);
      
      if (questions.length === 0) {
        console.log(`  No questions found on page ${pageNum}`);
        continue;
      }
      
      console.log(`  Found ${questions.length} questions: ${questions.map(q => q.questionNumber).join(', ')}`);
      
      // Extract each question
      for (const q of questions) {
        // Skip if question number is out of expected range
        if (q.questionNumber < 1 || q.questionNumber > info.count) {
          console.log(`  Skipping question ${q.questionNumber} (out of range)`);
          continue;
        }
        
        const outputPath = path.join(OUTPUT_DIR, `${year}_q${q.questionNumber}.png`);
        
        const success = await extractQuestionFromPage(pageImagePath, q, outputPath);
        
        if (success) {
          allExtractions[year][q.questionNumber] = outputPath;
          console.log(`  ✓ Extracted question ${q.questionNumber}${q.hasDiagram ? ' (has diagram)' : ''}`);
        } else {
          console.log(`  ✗ Failed to extract question ${q.questionNumber}`);
        }
      }
      
      // Small delay between pages
      await new Promise(resolve => setTimeout(resolve, 500));
    }
  }
  
  // Summary
  console.log('\n\n=== EXTRACTION SUMMARY ===');
  for (const [year, questions] of Object.entries(allExtractions)) {
    const count = Object.keys(questions).length;
    const expected = YEAR_PAGE_MAP[year].count;
    console.log(`${year}: ${count}/${expected} questions extracted`);
  }
  
  // Save extraction map
  const mapPath = path.join(OUTPUT_DIR, 'extraction_map.json');
  fs.writeFileSync(mapPath, JSON.stringify(allExtractions, null, 2));
  console.log(`\nExtraction map saved to: ${mapPath}`);
}

main().catch(console.error);
